## Model A指標計算

In [14]:
#Lets calculate Precision, Recall and F1 score for label 0 and 1
#For Label 0
tp = 853
fp = 341
fn = 576

precision_0  = tp / (tp + fp)
recall_0     = tp / (tp + fn)
f1_score_0   = 2*( precision_0 * recall_0)/(precision_0 + recall_0)

print('precision, recall and f1-score for label 真')
print('The precision for label 真 is: {0:.4f}'.format(precision_0))
print('The recall for label 真 is: {0:.4f}'.format(recall_0))
print('The f1-score for label 真 is: {0:.4f}'.format(f1_score_0))
print('\n')

#For Label 1 

tp = 7230
fp = 576
fn = 341

precision_1  = tp / (tp + fp)
recall_1     = tp / (tp + fn)
f1_score_1   = 2*( precision_1 * recall_1)/(precision_1 + recall_1)

print('precision, recall and f1-score for label 假')
print('The precision for label 假 is: {0:.4f}'.format(precision_1))
print('The recall for label 假 is: {0:.4f}'.format(recall_1))
print('The f1-score for label 假 is: {0:.4}'.format(f1_score_1))
print('\n')

print('precision, recall and f1-score for label Average')
print('The precision for label Average is: {0:.4f}'.format((precision_0+precision_1)/2))
print('The recall for label Average is: {0:.4f}'.format((recall_0 + recall_1) / 2))
print('The f1-score for label Average is: {0:.4f}'.format((f1_score_0 + f1_score_1) / 2))
print('\n')


precision, recall and f1-score for label 真
The precision for label 真 is: 0.7144
The recall for label 真 is: 0.5969
The f1-score for label 真 is: 0.6504


precision, recall and f1-score for label 假
The precision for label 假 is: 0.9262
The recall for label 假 is: 0.9550
The f1-score for label 假 is: 0.9404


precision, recall and f1-score for label Average
The precision for label Average is: 0.8203
The recall for label Average is: 0.7759
The f1-score for label Average is: 0.7954




## Model B指標計算

In [12]:
#Lets calculate Precision, Recall and F1 score for label 0 and 1
#For Label 0
tp = 846
fp = 316
fn = 583

precision_0  = tp / (tp + fp)
recall_0     = tp / (tp + fn)
f1_score_0   = 2*( precision_0 * recall_0)/(precision_0 + recall_0)

print('precision, recall and f1-score for label 真')
print('The precision for label 真 is: {0:.4f}'.format(precision_0))
print('The recall for label 真 is: {0:.4f}'.format(recall_0))
print('The f1-score for label 真 is: {0:.4f}'.format(f1_score_0))
print('\n')

#For Label 1 

tp = 7255
fp = 583
fn = 316

precision_1  = tp / (tp + fp)
recall_1     = tp / (tp + fn)
f1_score_1   = 2*( precision_1 * recall_1)/(precision_1 + recall_1)

print('precision, recall and f1-score for label 假')
print('The precision for label 假 is: {0:.4f}'.format(precision_1))
print('The recall for label 假 is: {0:.4f}'.format(recall_1))
print('The f1-score for label 假 is: {0:.4f}'.format(f1_score_1))
print('\n')

print('precision, recall and f1-score for label Average')
print('The precision for label Average is: {0:.4f}'.format((precision_0+precision_1)/2))
print('The recall for label Average is: {0:.4f}'.format((recall_0 + recall_1) / 2))
print('The f1-score for label Average is: {0:.4f}'.format((f1_score_0 + f1_score_1) / 2))
print('\n')


precision, recall and f1-score for label 真
The precision for label 真 is: 0.7281
The recall for label 真 is: 0.5920
The f1-score for label 真 is: 0.6530


precision, recall and f1-score for label 假
The precision for label 假 is: 0.9256
The recall for label 假 is: 0.9583
The f1-score for label 假 is: 0.9417


precision, recall and f1-score for label Average
The precision for label Average is: 0.8268
The recall for label Average is: 0.7751
The f1-score for label Average is: 0.7973




## 兩模型比較
#### Model A:
##### precision, recall and f1-score for label Average 
##### The precision for label Average is: 0.8203
##### The recall for label Average is: 0.7759
##### The f1-score for label Average is: 0.7954

#### Model B:
##### precision, recall and f1-score for label Average
##### The precision for label Average is: 0.8268
##### The recall for label Average is: 0.7751
##### The f1-score for label Average is: 0.7973

#### Ans：
Precision和Recall同時關注的都是True Positive（潛在分子），但是角度不一樣，
Precision看的是在預測正向的情形下，實際的精準度是多少，叫考慮的是模型的預測能力，
而Recall則是看在實際情況下為正向的狀況下，預測能找出多少實際正向的答案，較考慮實際正向的預測能力，
今天此題目為訂房銷售，則代表了Recall指標相對而言比Precision指標來得重要，因為我們會比較在意的是我們在所有客戶中，能夠找出最多潛在會訂房的客戶，讓利益最大化.
##### 因此從recall指標中可以看出ModelA 0.7759 > Model B 0.7751 因此我會選擇ModelA當作本情境的使用模型